packages/cardano-services/src/Program/utils.ts

```diff
@@ -0,0 +1,249 @@
+  dnsResolver: DnsResolver,
+  serviceName: string
+): Promise<TxSubmitProvider> => {
+  const { name, port } = await dnsResolver(serviceName!);
+  let ogmiosProvider = ogmiosTxSubmitProvider({ host: name, port });
+
+  return new Proxy<TxSubmitProvider>({} as TxSubmitProvider, {
+    get(_, prop) {
+      if (prop === 'then') return;
+      if (prop === 'submitTx') {
+        return (args: Uint8Array) =>
+          ogmiosProvider.submitTx(args).catch(async (error) => {
+            if (error instanceof WebSocketClosed && error.message === 'WebSocket is closed') {
+              const record = await dnsResolver(serviceName!);
+              ogmiosProvider = ogmiosTxSubmitProvider({ host: record.name, port: record.port });
+              return await ogmiosProvider.submitTx(args);
+            }
+          });
+      }
+      // Bind if it is a function, no intercept operations
+      if (typeof ogmiosProvider[prop as keyof TxSubmitProvider] === 'function') {
+        const method = ogmiosProvider[prop as keyof TxSubmitProvider] as any;
+        return method.bind(ogmiosProvider);
+      }
+    }
+  });
+};
+
+export const getOgmiosTxSubmitProvider = async (
+  dnsResolver: DnsResolver,
+  options?: CommonProgramOptions
+): Promise<TxSubmitProvider> => {
+  if (options?.ogmiosSrvServiceName)
+    return ogmiosTxSubmitProviderWithDiscovery(dnsResolver, options.ogmiosSrvServiceName);
+  if (options?.ogmiosUrl) return ogmiosTxSubmitProvider(urlToConnectionConfig(options?.ogmiosUrl));
+  throw new MissingProgramOption(ServiceNames.TxSubmit, [
+    CommonOptionDescriptions.OgmiosUrl,
+    CommonOptionDescriptions.OgmiosSrvServiceName
+  ]);
+};
+
+export const srvRecordToRabbitmqURL = ({ name, port }: SrvRecord) => new URL(`amqp://${name}:${port}`);
+
+/**
+ * Creates a extended RabbitMqTxSubmitProvider instance :
+ * - use passed srv service name in order to resolve the port
+ * - make dealing with failovers (re-resolving the port) opaque
+ * - use exponential backoff retry internally with default timeout and factor
+ * - intercept 'submitTx' operation and handle connection errors runtime
+ * - all other operations are bind to pool object withoud modifications
+ *
+ * @returns RabbitMqTxSubmitProvider instance
+ */
+export const rabbitMqTxSubmitProviderWithDiscovery = async (
+  dnsResolver: DnsResolver,
+  serviceName: string
+): Promise<RabbitMqTxSubmitProvider> => {
+  const record = await dnsResolver(serviceName!);
+  let rabbitmqProvider = new RabbitMqTxSubmitProvider({
+    rabbitmqUrl: srvRecordToRabbitmqURL(record)
+  });
+
+  return new Proxy<RabbitMqTxSubmitProvider>({} as RabbitMqTxSubmitProvider, {
+    get(_, prop) {
+      if (prop === 'then') return;
+      if (prop === 'submitTx') {
+        return (args: Uint8Array) =>
+          rabbitmqProvider.submitTx(args).catch(async (error) => {
+            if (error instanceof ProviderError && error.innerError === ProviderFailure.ConnectionFailure) {
```

Comment:
Really nice implementation!!

The only thing to note: in case the condition in this `if` is not matched, the error should be propagated to the caller.
I'm not expert with this construct: it could be it is propagated implicitly... I'm simply rising the flag as I can't understand how (since I don't see this explicitly).

Reply:
Good catch @iccicci , the missing line of code is introduced here https://github.com/input-output-hk/cardano-js-sdk/pull/309/commits/cd4da53c23624baa6b7bf944ec079ae9ed82fac1 : in case the error doesn't match the condition for connection error within the `Proxy` -> re-throw the error unchanged which propagates the error to the caller.

Unfortunately, I can not see currently how this case could be unit tested...
With service discovery, we use `Proxy` abstraction over the dependency services where in case of connection error runtime, we always try to [re-connect to a new instance](https://github.com/input-output-hk/cardano-js-sdk/blob/cd4da53c23624baa6b7bf944ec079ae9ed82fac1/packages/cardano-services/src/Program/utils.ts#L95-L99) internally of given dependency (_postgres,rabbitmq or ogmios_), or in this case: any other error type to be propagated unchanged. The main issue is with mocking such behavior with the underlying dependencies. (`pg.Pool`, `rabbitmqTxProvider` , `ogmiosTcProvider`)

As I am aware, simulating such behavior (killing a given healthy and running instance and expecting to recover/restart in some way) is associated with chaos testing with appropriate tools. Any suggestions are welcomed!

Reply:
I could suggest to start an ogmios mock like [this one](https://github.com/input-output-hk/cardano-js-sdk/blob/2cd240513a4632da7c6fef9274e65d025e803048/packages/rabbitmq/test/TxSubmitWorker.test.ts#L152).
This should be enough to test `ogmiosTxSubmitProviderWithDiscovery` while to test `rabbitMqTxSubmitProviderWithDiscovery` a txWorker would be required as well (while running `cardano-services` tests are running a RabbitMQ server is running as well inside docker).

Reply:
The failovers cases are covered under https://github.com/input-output-hk/cardano-js-sdk/pull/309/commits/87b50012fd1e88e722358e8f89bf17c797db9abc